tools/travis.sh

```diff
@@ -1,91 +1,23 @@
 #! /bin/bash
-#
+# This script attempts to install all upstream dependencies 
+# using OCaml 4.07 in a Docker container. This must be run 
+# from the root of the xs-opam repository.
+# 
 
-set -ex
+set -e
 
-# override xenctrl.dummy to have a full-fledged opam repository
-if [ ! "${COMPILE_ALL}" = 1 ]; then
-    sudo sh tools/into_repo.sh
-fi
+IMG='ocaml/opam2:debian-9-ocaml-4.06'
 
+docker pull $IMG
+docker run --rm -iv $PWD:/mnt $IMG <<'EOF'
+set -e
 
-pkg()
-{
-    pushd packages > /dev/null
-    find  $@\
-      -maxdepth 1 -mindepth 1 -type d \
-      | awk -F/ '{print $NF}'
-    popd > /dev/null
-}
+opam repo remove --all default
+opam repo add xs-opam file:///mnt
 
-enable_centos_virt_xen()
-{
-    case "${DISTRO}" in
-        centos*)
-            sudo cp /xs-opam/tools/CentOS-Xen.repo /etc/yum.repos.d/
-            sudo cp /xs-opam/tools/RPM-GPG-KEY-CentOS-SIG-Virtualization /etc/pki/rpm-gpg/
-            ;;
-        *) echo "Nothing to be done for DISTRO=${DISTRO};"
-            ;;
-    esac
-}
+cd /mnt
+opam depext -iy xs-toolstack
+EOF
 
-update_distro_packages()
-{
-    case "${DISTRO}" in
-        centos*)
-            sudo yum install -y epel-release
-            ;;
 
-        debian*)
-            sudo apt-get update -y
-            ;;
-        *)
-            echo "unknown DISTRO=${DISTRO}; nothing to do"
-            ;;
-    esac
-}
 
-if [ "${COMPILE_ALL}" = 1 ]; then
-    UPSTREAM="$(pkg upstream upstream-extra)"
-    XS="$(pkg xs xs-extra | grep -v xenctrl.dummy)"
-else
-    UPSTREAM="$(pkg upstream)"
-    XS="$(pkg xs)"
-fi
-
-enable_centos_virt_xen
-
-if [ "${SAFE_STRING}" = 0 -a "${OCAML_VERSION:0:4}" = "4.06" ]; then
-   opam switch create 4.06.0+default-unsafe-string
-   eval $(opam config env)
-fi
-
-if [ ! "${BASE_REMOTE}" = "" ]; then
-    opam repo remove --all default
-    opam repo add base "$BASE_REMOTE"
-fi
-
-if [ ! "${EXTRA_REMOTES}" = "" ]; then
-    opam repo add extra "$EXTRA_REMOTES"
-fi
-
-if [ "${OPAM_LINT}" = 1 ]; then
-    find packages -iname opam -print | xargs -n 1 opam lint
-elif [ "${CHECK_UNUSED}" = 1 ]; then
-    opam install -y --fake $XS $(pkg xs-extra-dummy)
-    # Fail if there are unused packages in $UPSTREAM:
-    AVAILABLE=$(pkg upstream | sed 's/\..*$//' | sort -u)
-    INSTALLED=$(opam list | grep -v \# | cut -d' ' -f1 | sort -u)
-    UNNEEDED=$(comm -23 <(echo "$AVAILABLE") <(echo "$INSTALLED"))
-    if [ -n "$UNNEEDED" ]; then
-        echo Unused packages in upstream/: $UNNEEDED
-        exit 1
-    fi
-elif [ "${CHECK_TEST_DEPENDENCIES}" = 1 ]; then
-    opam install -t -y --fake $XS
```

Comment:
We could keep these dependency checks, but in a separate script. I agree that there were too many of them. Perhaps just one script and Travis matrix entry that does all the dependency, and unused package checks?

Or we could have these checks done in the `tools/opam-updates.py` script that we run from a cronjob, what do you think?

About the docker container: I am using it here: https://github.com/edwintorok/xs-opam-dev, but if it isn't generally useful I could move building the base of it to my repo as well. (although not sure about `tools/opam-updates.py` it might need the container). Even if we keep the docker container we could use a simplified form of it and build just one combination (say CentOS-7 cause thats what we use internally) instead of being generic for any distribution.

Reply:
I understand that `opam-updates` uses the container but I would propose we add building the container back in later.